In [38]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, OneHotEncoder


# Import iterative Imputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
# To preprocess the data
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder, OrdinalEncoder, OneHotEncoder
from sklearn.impute import SimpleImputer, KNNImputer

# Import iterative Imputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# For Machine Learning
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score

# Metrics
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, mean_absolute_error, precision_score, mean_squared_error, r2_score

# For Classification\Regression Tasks
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingRegressor, GradientBoostingClassifier
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split, GridSearchCV, LeaveOneOut, RandomizedSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split


from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
# For Machine Learning
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score

# Metrics
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, mean_absolute_error, precision_score, mean_squared_error, r2_score

import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense


In [39]:
df = pd.read_csv("H:\\MY\\Computer Science\\Python Codes\\ML Pipeline\\Final Data of Salary Dataset.csv")
df

,Unnamed: 0,Age,Gender,Education Level,Years of Experience,Salary,Quantile_X,Quantile_Y,Quantile_Z
0,0,32.0,Male,Bachelor's,5.0,103000.0,0.061513,-0.257495,-0.121416
1,1,28.0,Female,Master's,3.0,65000.0,-0.604448,-0.700711,-0.723343
2,2,45.0,NaN,PhD,15.4,150000.0,1.394512,1.207012,0.551076
3,3,36.0,NaN,Bachelor's,7.0,60000.0,0.556927,0.056486,-0.837338
4,4,52.0,Male,Master's,8.0,200000.0,5.199338,0.208505,1.994971
...,...,...,...,...,...,...,...,...,...
6545,6698,49.0,Female,PhD,20.0,200000.0,1.942703,1.895259,1.994971
6546,6699,32.0,Male,High School,3.0,50000.0,0.061513,-0.700711,-1.181921
6547,6700,30.0,Female,Bachelor's,4.0,55000.0,-0.232930,-0.445919,-0.999986
6548,6701,46.0,Male,Master's,14.0,140000.0,1.533612,1.025123,0.407442


In [40]:
df.drop(columns=["Unnamed: 0", "Quantile_X",	"Quantile_Y",	"Quantile_Z"], inplace=True)
df

,Age,Gender,Education Level,Years of Experience,Salary
0,32.0,Male,Bachelor's,5.0,103000.0
1,28.0,Female,Master's,3.0,65000.0
2,45.0,NaN,PhD,15.4,150000.0
3,36.0,NaN,Bachelor's,7.0,60000.0
4,52.0,Male,Master's,8.0,200000.0
...,...,...,...,...,...
6545,49.0,Female,PhD,20.0,200000.0
6546,32.0,Male,High School,3.0,50000.0
6547,30.0,Female,Bachelor's,4.0,55000.0
6548,46.0,Male,Master's,14.0,140000.0


In [41]:
df.isnull().sum()

Age                    0
Gender                 3
Education Level        3
Years of Experience    0
Salary                 0
dtype: int64

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6550 entries, 0 to 6549
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  6550 non-null   float64
 1   Gender               6547 non-null   object 
 2   Education Level      6547 non-null   object 
 3   Years of Experience  6550 non-null   float64
 4   Salary               6550 non-null   float64
dtypes: float64(3), object(2)
memory usage: 256.0+ KB


In [43]:
le = LabelEncoder()
df["Gender"] = le.fit_transform(df["Gender"])

In [44]:
le = LabelEncoder()
df["Education Level"] = le.fit_transform(df["Education Level"])

In [45]:
from sklearn.impute import KNNImputer
impute = KNNImputer()
df["Gender"] = impute.fit_transform(df[["Gender"]])

In [46]:
from sklearn.impute import KNNImputer
impute = KNNImputer()
df["Education Level"] = impute.fit_transform(df[["Education Level"]])

In [47]:
df

,Age,Gender,Education Level,Years of Experience,Salary
0,32.0,1.0,0.0,5.0,103000.0
1,28.0,0.0,2.0,3.0,65000.0
2,45.0,3.0,3.0,15.4,150000.0
3,36.0,3.0,0.0,7.0,60000.0
4,52.0,1.0,2.0,8.0,200000.0
...,...,...,...,...,...
6545,49.0,0.0,3.0,20.0,200000.0
6546,32.0,1.0,1.0,3.0,50000.0
6547,30.0,0.0,0.0,4.0,55000.0
6548,46.0,1.0,2.0,14.0,140000.0


In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6550 entries, 0 to 6549
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  6550 non-null   float64
 1   Gender               6550 non-null   float64
 2   Education Level      6550 non-null   float64
 3   Years of Experience  6550 non-null   float64
 4   Salary               6550 non-null   float64
dtypes: float64(5)
memory usage: 256.0 KB


In [49]:
df.shape


(6550, 5)

In [50]:
print(df["Salary"].min())
print(df["Salary"].max())

350.0
250000.0


In [51]:
X = df.drop("Salary", axis=1)
y = df["Salary"]

X_train, X_test, y_train, y_test =train_test_split(X,y, test_size=0.2, random_state=42)

import numpy as np

# Convert to numeric if possible
X_train = np.array(X_train, dtype=np.float32)
X_test  = np.array(X_test, dtype=np.float32)
y_train = np.array(y_train, dtype=np.float32)
y_test  = np.array(y_test, dtype=np.float32)



In [52]:
model = models.Sequential([
    
    layers.Dense(512, activation='relu', input_shape=(4,)),
    layers.Dense(256, activation='relu'),
    layers.Dense(128, activation='relu'),
    layers.Dense(64, activation='relu'),
    layers.Dense(32, activation='relu'),
    layers.Dense(16, activation='relu'),
    layers.Dense(1, activation='linear'),
])

model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_16 (Dense)            (None, 512)               2560      
                                                                 
 dense_17 (Dense)            (None, 256)               131328    
                                                                 
 dense_18 (Dense)            (None, 128)               32896     
                                                                 
 dense_19 (Dense)            (None, 64)                8256      
                                                                 
 dense_20 (Dense)            (None, 32)                2080      
                                                                 
 dense_21 (Dense)            (None, 16)                528       
                                                                 
 dense_22 (Dense)            (None, 1)                

In [53]:
model.compile(optimizer='adam',
              loss='mse',
              metrics=['mae'])

In [63]:
history= model.fit(X_train, y_train, epochs= 180,
                 validation_data=(X_test, y_test))

Epoch 1/180
164/164 [==============================] - 2s 14ms/step - loss: 361675360.0000 - mae: 14104.6533 - val_loss: 439481760.0000 - val_mae: 14973.3516
Epoch 2/180
164/164 [==============================] - 2s 12ms/step - loss: 363820672.0000 - mae: 14055.1250 - val_loss: 490306336.0000 - val_mae: 16752.8926
Epoch 3/180
164/164 [==============================] - 2s 11ms/step - loss: 376869664.0000 - mae: 14401.8057 - val_loss: 426833312.0000 - val_mae: 15513.3926
Epoch 4/180
164/164 [==============================] - 2s 11ms/step - loss: 377825696.0000 - mae: 14527.3818 - val_loss: 412185408.0000 - val_mae: 14692.7500
Epoch 5/180
164/164 [==============================] - 2s 9ms/step - loss: 362351008.0000 - mae: 14122.6104 - val_loss: 411654208.0000 - val_mae: 14969.0732
Epoch 6/180
164/164 [==============================] - 2s 11ms/step - loss: 366948960.0000 - mae: 14250.8379 - val_loss: 425176800.0000 - val_mae: 15066.6338
Epoch 7/180
164/164 [==============================] 

In [64]:
# Evaluate
loss, mae = model.evaluate(X_test, y_test, verbose=0)
print(f"Test MAE: {mae:.2f}")

Test MAE: 15105.25


2. Relative to Salary Range (350 – 250,000)
Salary range ≈ 249,650
Normalized MAE = 23,929 ÷ 249,650 ≈ 9.6% of the full range


3. Relative to Mean Salary (~95,000)
Normalized MAE = 23,929 ÷ 95,000 ≈ 25% of average salary
This is high → predictions are off by about a quarter of the typical salary.

4. Interpretation
<10% of mean salary → excellent
10–15% → acceptable
>20% → needs improvement

5. Final Target 
For this dataset:
Aim for MAE = 10k–15k
That would be ~10–15% of mean salary, which is a much better performance.

In [65]:
model.save("Salary Deep Model.H5")

INFO:tensorflow:Assets written to: Salary Deep Model.H5\assets


INFO:tensorflow:Assets written to: Salary Deep Model.H5\assets
